### PR TITLE
Added soft limit to open scroll contexts #25244

### LINKED ISSUE #25244: Add a softlimit to the number of open scroll contexts per node
Today if a user creates a large number of scroll contexts without clearing them using a scroll ID we go and clean things up once a timeout is reached. This has several problems:

 * the user doesn't necessarily realize an API call to release resources is necessary
 * the node piles up open readers behind the scenes that can have a big impact on disk space and monitoring.
 * the node might go OOM due to too many open resources
 * the node might suffer from too many open files 

we do protect against some of these but it would be nicer to reject scroll requests / new sticky context creation if there are too many of them. This change is rather a low hanging fruit but I suspect a big impact if a users has the related problems. A rejected scroll search would immediately be visible to the user.

### PR DESCRIPTION
**Pull request by joaofcmb**
Missing updated documentation. Will be added later.

Since development had already begun on this issue before the suggestions made by @geekpete , we thought it would be best to first make a pull request for the initial situation.

Closes #25244

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -696,6 +710,7 @@ public boolean freeContext(long id) {
                 assert context.refCount() > 0 : " refCount must be > 0: " + context.refCount();
                 context.indexShard().getSearchOperationListener().onFreeContext(context);
                 if (context.scrollContext() != null) {
+                    openScrollContexts.decrementAndGet();
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java]: we should also check the scroll contexts in `freeAllContextForIndex` above ?

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 2]
```diff
@@ -145,6 +146,9 @@
     public static final Setting<Boolean> DEFAULT_ALLOW_PARTIAL_SEARCH_RESULTS =
             Setting.boolSetting("search.default_allow_partial_results", true, Property.Dynamic, Property.NodeScope);
 
+    public static final Setting<Integer> MAX_OPEN_SCROLL_CONTEXT =
+        Setting.intSetting("search.max_open_scroll_context", 150, 0, Property.NodeScope);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java]: This setting should be dynamic, you should extract the value in the SearchService ctr and register an update consumer. DEFAULT_ALLOW_PARTIAL_SEARCH_RESULTS is a good example of how to do it.

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @joaofcmb , the pr looks good. I left some comments and questions.

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 1 above)
Comment (joaofcmb) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: Doesn't `freeAllContextForIndex` simply call `freeContext` ? Am I supposed to do any extra processing here?

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: Sorry my bad, it does call `freeContext`

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 2 above)
Comment (joaofcmb) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: Should be dynamic now. Please review the changes.

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 3]
```diff
@@ -145,6 +146,9 @@
     public static final Setting<Boolean> DEFAULT_ALLOW_PARTIAL_SEARCH_RESULTS =
             Setting.boolSetting("search.default_allow_partial_results", true, Property.Dynamic, Property.NodeScope);
 
+    public static final Setting<Integer> MAX_OPEN_SCROLL_CONTEXT =
+        Setting.intSetting("search.max_open_scroll_context", 150, 0, Property.Dynamic, Property.NodeScope);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java]: Let's start at `500` ? That's the number we agreed on in https://github.com/elastic/elasticsearch/pull/29252.

**Review by jimczi — COMMENTED** [review]
Thanks @joaofcmb , the code looks good. Can you add a note in the docs ?

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 3 above)
Comment (joaofcmb) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: Sure, i'll change it while I update the docs

**Comment by jimczi** [comment]
@elasticmachine test this please

**Review by jimczi — COMMENTED** [review]
Thanks @joaofcmb , the changes look good. The tests are running on CI now and I'll merge if they succeed.

**Comment by jimczi** [comment]
@elasticmachine test this please

**Review by jimczi — APPROVED** [review]
Thanks @joaofcmb . I'll backport to 6.x but will set the default to unlimited in this version.

**Pull request merged**